constraint-solver/src/solver/constraint_splitter.rs

```diff
@@ -0,0 +1,565 @@
+    let expr = expr - &GroupedExpression::from_number(expr_shift);
+    let rest = rest - GroupedExpression::from_number(rest_shift);
+    let constant = constant + expr_shift + smallest_coeff * rest_shift;
+
+    // rc(expr): [0,max_expr]
+    // rc(rest): [0,max_rest]
+    // If max_expr + k * max_rest < P, then we can translate the equation to the natural numbers:
+    // expr + k * rest = (-constant) % modulus
+
+    // TODO this does not work for large fields.
+    let expr_rc = expr.range_constraint(range_constraints);
+    let rest_rc = rest.range_constraint(range_constraints);
+    if !expr_rc.range().0.is_zero() || !rest_rc.range().0.is_zero() {
+        // The range constraints were unconstrained to begin with.
+        return None;
+    }
+
+    let max_expr = expr_rc.range().1;
+    let max_rest = rest_rc.range().1;
+
+    // Evaluate `expr + smallest_coeff * rest` for the largest possible value
+    // and see if it wraps around in the field.
+    if max_expr.to_arbitrary_integer()
+        + smallest_coeff.to_arbitrary_integer() * max_rest.to_arbitrary_integer()
+        >= T::FieldType::modulus().to_arbitrary_integer()
+    {
+        return None;
+    }
+    // It does not wrap around, so we know that the equation can be translated to the
+    // natural numbers:
+    // expr + smallest_coeff * rest = (-constant) % modulus
+
+    // Next, we apply `x -> x % smallest_coeff` to both sides of the equation to get
+    // expr % smallest_coeff = ((-constant) % modulus) % smallest_coeff
+    // Note that at this point, we only get an implication, not an equivalence,
+    // but if the range constraints of `expr` only allow a unique solution,
+    // we it still holds unconditionally.
+
+    if max_expr.to_integer() >= smallest_coeff.to_integer() + smallest_coeff.to_integer() {
+        // In this case, there are always at least two solutions (ignoring masks and other
+        // constraints).
+        return None;
+    }
+
+    let rhs = T::FieldType::from(
+        (-constant).to_integer().try_into_u64().unwrap()
+            % smallest_coeff.to_integer().try_into_u64().unwrap(),
+    );
+
+    // Now we try `rhs`, `rhs + smallest_coeff`, `rhs + 2 * smallest_coeff`, ...
+    // But because of the check above, we can stop at `2 * smallest_coeff`.
+    let solution = (0..=2)
+        .map(|i| rhs + T::FieldType::from(i) * smallest_coeff)
+        .filter(|candidate| expr_rc.allows_value(*candidate))
+        .exactly_one()
+        .ok()?;
+
+    // There is exactly one solution to the equation
+    // `expr % smallest_coeff = rhs`.
+    // Now we translate this back to the un-shifted expression.
+    Some(solution + expr_shift)
+}
+
+fn recombine_rest<T: RuntimeConstant + Display, V: Clone + Ord + Display>(
+    components: Vec<Component<T, V>>,
+    constant: T::FieldType,
+) -> AlgebraicConstraint<GroupedExpression<T, V>> {
+    let remaining = components
+        .into_iter()
+        .filter(|comp| !comp.is_zero())
```

Comment:
Is this required? Wouldn't the zero ones get removed somewhere else? Or is it for performance?